src/drivers/dshot/DShot.cpp

```diff
@@ -239,18 +251,74 @@ void DShot::handle_new_telemetry_data(const int telemetry_index, const DShotTele
 		esc_status.esc_connectiontype = esc_status_s::ESC_CONNECTION_TYPE_DSHOT;
 		esc_status.esc_count = _telemetry->handler.numMotors();
 		++esc_status.counter;
-		// FIXME: mark all ESC's as online, otherwise commander complains even for a single dropout
-		esc_status.esc_online_flags = (1 << esc_status.esc_count) - 1;
-		esc_status.esc_armed_flags = (1 << esc_status.esc_count) - 1;
-
-		_telemetry->esc_status_pub.update();
 
-		// reset esc data (in case a motor times out, so we won't send stale data)
-		memset(&esc_status.esc, 0, sizeof(_telemetry->esc_status_pub.get().esc));
-		esc_status.esc_online_flags = 0;
+		ret = 1; // Indicate we wrapped, so we publish data
 	}
 
 	_telemetry->last_telemetry_index = telemetry_index;
+
+	return ret;
+}
+
+void DShot::publish_esc_status(void)
+{
+	esc_status_s &esc_status = _telemetry->esc_status_pub.get();
+	uint8_t channel;
+
+	// clear data of the esc that are offline
+	for (channel = 0; (channel < _telemetry->last_telemetry_index); channel++) {
+		if ((esc_status.esc_online_flags & (1 << channel)) == 0) {
+			memset(&esc_status.esc[channel], 0, sizeof(struct esc_report_s));
+		}
+	}
+
+	// FIXME: mark all UART Telemetry ESC's as online, otherwise commander complains even for a single dropout
+	esc_status.esc_online_flags = (1 << esc_status.esc_count) - 1;
+	esc_status.esc_armed_flags = (1 << esc_status.esc_count) - 1;
+
+	if (_bdshot) {
+		esc_status.esc_online_flags |= _output_mask;
+		esc_status.esc_armed_flags |= _output_mask;
+		esc_status.esc_count = _dshot_esc_count;
+
+		for (channel = 0; (channel < 8); channel++) {
```

Comment:
The telemetry_index (= index to esc_status) does not necessarily correspond to the channel. There might be gaps if some motors are disabled.
You can check the method `update_telemetry_num_motors`. I think we should apply the same here.

Reply:
Thanks, the `8` was an hard-coded leftover piece also `_dshot_esc_count` was redundant replaced by `_telemetry->handler.numMotors();` instead.